vmapi crashes when sdc-migrate -n given name instead of UUID

If you run "sdc-migrate -n somename UUID" where somename is a text string rather than a UUID of a compute node, sdc-migrate returns "curl: (52) Empty reply from server" and vmapi crashes:

```
[2019-11-01T11:24:40.656Z]  INFO: moray/19029 on 2c2752be-2211-473d-8105-5de766bbc0a6: getVmMigrations
    filter: (&(|(source_server_uuid=00000000-0000-0000-0000-0cc47aa30dba)(&(target_server_uuid=00000000-0000-0000-0000-0cc47aa30dba)(state=successful)))(vm_uuid=0cd07440-fedd-40fb-a9a2-9cba5647f500))
Uncaught AssertionError: req.params.override_server_uuid (uuid) is required

FROM
_toss (/opt/smartdc/vmapi/node_modules/assert-plus/assert.js:22:5)
Function.out.(anonymous function) [as uuid] (/opt/smartdc/vmapi/node_modules/assert-plus/assert.js:122:17)
start_create_new_record (/opt/smartdc/vmapi/lib/vm-migration/migrate.js:549:24)
Immediate.<anonymous> (/opt/smartdc/vmapi/node_modules/vasync/lib/vasync.js:875:12)
runCallback (timers.js:672:20)
tryOnImmediate (timers.js:645:5)
processImmediate [as _immediateCallback] (timers.js:617:5)
[ Nov  1 11:24:41 Stopping because all processes in service exited. ]
[ Nov  1 11:24:41 Executing stop method (:kill). ]
[ Nov  1 11:24:41 Executing start method ("/opt/smartdc/vmapi/build/node/bin/node --abort_on_uncaught_exception /opt/smartdc/vmapi/server.js &"). ]
[ Nov  1 11:24:41 Method "start" exited with status 0. ]
```

--- a/lib/vm-migration/migrate.js
+++ b/lib/vm-migration/migrate.js
@@ -414,6 +414,28 @@ function migrateVm(req, res, next) {
             }
 
             if (action === 'begin' || action === 'estimate') {
+                // Validate override params.
+                if (req.params.override_server_uuid &&
+                        !common.validUUID(req.params.override_server_uuid)) {
+                    cb(new errors.ValidationFailedError(
+                        'Invalid parameters: override_server_uuid (' +
+                        req.params.override_server_uuid +
+                        ') is not a valid UUID',
+                        [ errors.invalidUuidErrorsElem('override_server_uuid') ]
+                        ));
+                    return;
+                }
+
+                if (req.params.override_uuid &&
+                        !common.validUUID(req.params.override_uuid)) {
+                    cb(new errors.ValidationFailedError(
+                        'Invalid parameters: override_uuid (' +
+                        req.params.override_uuid + ') is not a valid UUID',
+                        [ errors.invalidUuidErrorsElem('override_uuid') ]
+                        ));
+                    return;
+                }
+
                 ctx.app = req.app;
                 validateMigrationBegin(vm, ctx, cb);
                 return;

--- a/lib/vm-migration/migrate.js
+++ b/lib/vm-migration/migrate.js
@@ -50,10 +50,26 @@ function validateMigrationBegin(vm, ctx, callback) {
     assert.optionalBool(ctx.app.options.userMigrationAllowed,
         'ctx.app.options.userMigrationAllowed');
     assert.optionalObject(ctx.migrationRecord, 'ctx.migrationRecord');
+    assert.object(ctx.req, 'ctx.req');
     assert.func(callback, 'callback');
     assert.optionalUuid(ctx.ownerUuid, 'ctx.ownerUuid');
 
+    var checkUuidProps = ['override_server_uuid', 'override_uuid'];
+    var i;
+    var p;
     var vmAllowed = false;
+
+    // Validate override params.
+    for (i = 0; i < checkUuidProps.length; i++) {
+        p = checkUuidProps[i];
+        if (ctx.req.params[p] && !common.validUUID(ctx.req.params[p])) {
+            callback(new errors.ValidationFailedError(util.format(
+                'Invalid parameters: %s (%s) is not a valid UUID',
+                p, ctx.req.params[p]), [ errors.invalidUuidErrorsElem(p) ]
+            ));
+            return;
+        }
+    }
 
     if (!ctx.ownerUuid) {
         // This is an operator initiated migration - always allow it.
@@ -414,29 +430,8 @@ function migrateVm(req, res, next) {
             }
 
             if (action === 'begin' || action === 'estimate') {
-                // Validate override params.
-                if (req.params.override_server_uuid &&
-                        !common.validUUID(req.params.override_server_uuid)) {
-                    cb(new errors.ValidationFailedError(
-                        'Invalid parameters: override_server_uuid (' +
-                        req.params.override_server_uuid +
-                        ') is not a valid UUID',
-                        [ errors.invalidUuidErrorsElem('override_server_uuid') ]
-                        ));
-                    return;
-                }
-
-                if (req.params.override_uuid &&
-                        !common.validUUID(req.params.override_uuid)) {
-                    cb(new errors.ValidationFailedError(
-                        'Invalid parameters: override_uuid (' +
-                        req.params.override_uuid + ') is not a valid UUID',
-                        [ errors.invalidUuidErrorsElem('override_uuid') ]
-                        ));
-                    return;
-                }
-
                 ctx.app = req.app;
+                ctx.req = req;
                 validateMigrationBegin(vm, ctx, cb);
                 return;
             }

--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "vmapi",
   "description": "VMs API",
-  "version": "9.8.16",
+  "version": "9.8.17",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {

--- a/test/lib/migration.js
+++ b/test/lib/migration.js
@@ -35,6 +35,28 @@ function getJobError(job) {
     return null;
 }
 
+function expectErrorProperties(t, err, expectedErr) {
+    assert.object(t, 't');
+    assert.object(err, 'err');
+    assert.object(expectedErr, 'expectedErr');
+
+    Object.keys(expectedErr).forEach(function (prop) {
+        if (!err.hasOwnProperty(prop)) {
+            t.ok(false, 'Error object is missing the "' + prop + '" field');
+            return;
+        }
+
+        if (err[prop] != expectedErr[prop]) {
+            t.ok(false, 'Error property "' + prop + '" is not as expected. ' +
+                'Expected "' + expectedErr[prop] + '" but got "' +
+                err[prop] + '"');
+            return;
+        }
+
+        t.ok(true, 'Error property "' + prop + '" is as expected');
+    });
+}
+
 function MigrationWatcher(client, vm_uuid) {
     this.client = client;
     this.vm_uuid = vm_uuid;
@@ -326,6 +348,46 @@ function TestMigrationCfg(test, cfg) {
             if (err) {
                 t.equal(err.statusCode, 409,
                     format('err.statusCode === 409, got %s', err.statusCode));
+            }
+            t.done();
+        });
+    };
+
+    test.bad_migrate_invalid_override_server_uuid = function (t) {
+        // No action.
+        client.post({
+            path: format('/vms/%s?action=migrate&migration_action=begin&' +
+                'override_server_uuid=foobar', sourceVm.uuid)
+        }, function onMigrateBadOverrideServerUuid(err, req, res, body) {
+            t.ok(err, 'expect error when an invalid server_uuid is supplied');
+            if (err && body) {
+                t.equal(err.statusCode, 409,
+                    format('err.statusCode === 409, got %s', err.statusCode));
+                expectErrorProperties(t, body, {
+                    code: 'ValidationFailed',
+                    message: 'Invalid parameters: override_server_uuid ' +
+                        '(foobar) is not a valid UUID'
+                });
+            }
+            t.done();
+        });
+    };
+
+    test.bad_migrate_invalid_override_uuid = function (t) {
+        // No action.
+        client.post({
+            path: format('/vms/%s?action=migrate&migration_action=begin&' +
+                'override_uuid=fishstick', sourceVm.uuid)
+        }, function onMigrateBadOverrideUuid(err, req, res, body) {
+            t.ok(err, 'expect error when an invalid override_uuid is supplied');
+            if (err && body) {
+                t.equal(err.statusCode, 409,
+                    format('err.statusCode === 409, got %s', err.statusCode));
+                expectErrorProperties(t, body, {
+                    code: 'ValidationFailed',
+                    message: 'Invalid parameters: override_uuid (fishstick) ' +
+                        'is not a valid UUID'
+                });
             }
             t.done();
         });

--- a/test/lib/migration.js
+++ b/test/lib/migration.js
@@ -46,7 +46,7 @@ function expectErrorProperties(t, err, expectedErr) {
             return;
         }
 
-        if (err[prop] != expectedErr[prop]) {
+        if (err[prop] !== expectedErr[prop]) {
             t.ok(false, 'Error property "' + prop + '" is not as expected. ' +
                 'Expected "' + expectedErr[prop] + '" but got "' +
                 err[prop] + '"');

